configuration-rendering-controller/README.md

```diff
@@ -0,0 +1,221 @@
+are all for the MC and they are all reconciled by the unique `app-operator`, which has RBAC rights to access the
+generated CMs and Secrets in any namespace.
+
+For workload cluster `app-operators` that might need access to configurations in the future, we could either give
+access to the namespace for CMs and Secrets that contains all rendered configs - if they would use the same -
+or render the workload cluster specific level to the given namespace. For one, the CR should be able to support that
+scope change. For another we do not have that implemented in the current system. The main issue here however is not
+a problem with this system, but with what we can share freely with customer workloads. For example, `shared-configs`
+is still kept private, because it contains - even if each encrypted with MC unique keys - minor Giant Swarm sensitive
+information. They are required by vintage to my knowledge, but they are there nevertheless and a security audit
+should be carried before we make that repository public.
+
+### Extending configuration management in the future
+
+Currently, we mostly utilise Catalog configuration for cluster apps. I think the reason for it is that it is not
+possible at the moment to utilise the configuration system when we deploy such workloads, so the configuration
+drifted into the catalogs, as the `cluster` catalog has well-defined scope for such apps.
+
+With easy access to configuration management, these catalog configs could simply drift into the config structure
+for the apps that need them on the MC specific level, and we could then simply get rid of them.
+
+## Implementation approaches
+
+### Framework
+
+The new controller should be written using [kubebuilder](https://book.kubebuilder.io/). It is on our roadmap to
+[Remove `operatorkit` as a dependency](https://github.com/giantswarm/roadmap/issues/3722) and this would be a great
+opportunity to get some hands-on experience with `kubebuilder`, as most of those operators that need and upgrade
+is owned by Team HoneyBadger.
+
+### Possibility to utilize Flux source-controller
+
+It is already implemented in `konfigure` to fetch the sources from Flux `source-controller` with `include`
+feature support [here](https://github.com/giantswarm/konfigure/pull/278).
+
+It would potentially be best to not depend on Flux either, but since we already have it implemented, and we would
+need to do something very similar to fetch and assemble the sources, we could cut this corner now and keep
+using Flux.
+
+There is potential to later implement custom DSL to describe how to merge configurations, which could make sense
+if we want to introduce many different layers in the future (catalogs, workload-cluster, app instance, etc. specific
+configs). For now, it makes sense to keep our current logic as is to reduce time to implement and friction.
+
+### Defining scope of configs to render
+
+TBD
```

Comment:
That's the interesting part. Here I mean by scope "which parts of the source config repos to apply". Let's try to split it into 2 sub-cases: what we want to support now and what we think might be useful in the future (to not block it by decisions we take now).

The only current scope is `Installation` and `App`. I think it's OK to keep it like that to get started. This means, we get exactly the same config for each app as long as we render it for the same installation. We can save the resulting CM&S into the `giantswarm` namespace and have a reasonable starting point, that also mimics how `collections` work now.

In the future, we want:
- WC scope (atlas has a use case for that already) - different config for each WC, but the same App. Currently, this can be generalised as "different config for different app instances". We can think about matching by app CR name (like a name regexp pattern, but that's fragile) or by some labels present on the App CR. 

- maybe org level, but I don't think I've seen a use case for that

Reply:
I am looking for some common expression language maybe, or ensuring extensibility by setting type to scopes, e.g:

```
scopes:
  - type: managment-cluster
    expression: *

# or

scopes:
  - type: workload-cluster
     expression: operations

# or

scopes:
  - type: app-instances
    expression: kyverno*|cert-manager|external-dns
```

In case of multiple scopes defined, its always additive. All configs that match any of the rules would be rendered.

Another approach would be to define a different CRD per kind of configuration. That is very flexible open for expansion. We can start with 1 for the installation use case we have now, just name it accordingly and later we can add more for catalogs, WCs and instances. With `kubebuilder` we can even have the go structs "inherit" the shared parts to avoid duplication.

Reply:
Detailed this in a new section.